### PR TITLE
Emit event from AddServer and RemoveServer

### PR DESCRIPTION
**Pull request by CorgiMan**
After putting the AddServer and RemoveServer back in, I forgot to let them emit the RingChanged event. I also put enhanced the tests to check for these events so this won't happen again.

@uber/ringpop

### CODE REVIEW

File: hashring/hashring.go [hunk 1]
```diff
@@ -64,7 +64,7 @@ func (r *HashRing) emit(event interface{}) {
 	}
 }
 
-// RegisterListener adds a listener that will be sent swim events.
+// RegisterListener adds a listener that will listens for HashRing events.
```
Comment (dansimau) [inline hashring/hashring.go]: Not sure about this. I actually think this comment was more accurate before. In any case, there is a typo in the updated comment.

File: hashring/hashring.go [hunk 2]
```diff
@@ -197,8 +195,8 @@ func (r *HashRing) addRemoveServersNoLock(add []string, remove []string) bool {
 	}
 
 	if changed {
-		r.emit(events.RingChangedEvent{add, remove})
 		r.computeChecksumNoLock()
+		r.emit(events.RingChangedEvent{add, remove})
```
Comment (dansimau) [inline hashring/hashring.go]: :+1:

File: hashring/hashring.go (hunk 1 above)
Comment (dansimau) [inline hashring/hashring.go]: Actually, scratch that. I just looked more closely at the events and you're right. A listener does receive all event types. My bad. Just the typo then: `s/listens/listen`! :smile:

File: hashring/hashring_test.go [hunk 3]
```diff
@@ -52,42 +52,45 @@ func (d *dummyListener) HandleEvent(event events.Event) {
...
-	ring.emit(struct{}{})
-	assert.Equal(t, 1, l.events, "expected one event to be emitted")
-}
 
-func TestAddServer(t *testing.T) {
-	ring := New(farm.Fingerprint32, 10)
 	ring.AddServer("server1")
 	ring.AddServer("server2")
-
+	assert.Equal(t, 4, l.EventCount())
```
Comment (dansimau) [inline hashring/hashring_test.go]: I know it's been mentioned before—and this is just a minor thing—but I think we should move to using generated mocks where possible, for the sake of ease and standardisation.

With the generated mock you could do:

```
// Set up mock event listener with expectations
mockEventListener := &mocks.EventListener{}
mockEventListener.On("HandleEvent", mock.AnythingOfType("RingChangedEvent")).Times(5)
ring.RegisterListener(mockEventListener)

// Your test code here ...

// Assert expectations
mockEventListener.AssertExpectations()    // Will complain if the RingChangedEvent wan't fired 5 times
```

An example is here in this PR:
https://github.com/uber/ringpop-go/blob/tests/swim/swim/ping_test.go#L85

Note that I don't think this is a blocker for landing the PR. Just a nice-to-have.

**Comment by dansimau** [comment]
Code lgtm :+1:

**Comment by thanodnl** [comment]
:+1:

File: hashring/hashring_test.go (hunk 3 above)
Comment (CorgiMan) [inline hashring/hashring_test.go]: I agree, but the mockEventListener is not yet in master right? We can merge the changes in there, or clean up when that lands.

**Comment by CorgiMan** [comment]
Thanks for the quick review boys

**Pull request merged**